shared-bindings/displayio/Bitmap.c

```diff
@@ -172,6 +172,115 @@ STATIC mp_obj_t bitmap_subscr(mp_obj_t self_in, mp_obj_t index_obj, mp_obj_t val
+        {MP_QSTR_x, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_y, MP_ARG_REQUIRED | MP_ARG_INT},
+        {MP_QSTR_source, MP_ARG_REQUIRED | MP_ARG_OBJ},
+        {MP_QSTR_x1, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        {MP_QSTR_y1, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} },
+        {MP_QSTR_x2, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} }, // None convert to source->width
+        {MP_QSTR_y2, MP_ARG_KW_ONLY | MP_ARG_OBJ, {.u_obj = mp_const_none} }, // None convert to source->height
+        {MP_QSTR_skip_index, MP_ARG_OBJ | MP_ARG_KW_ONLY, {.u_obj=mp_const_none} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args - 1, pos_args + 1, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    displayio_bitmap_t *self = MP_OBJ_TO_PTR(pos_args[0]); //*******
+
+    int16_t x = args[ARG_x].u_int;
+    int16_t y = args[ARG_y].u_int;
+
+    displayio_bitmap_t *source = MP_OBJ_TO_PTR(args[ARG_source].u_obj);
+
+    int16_t x1, y1;
+    // if x1 or y1 is None, then set as the zero-point of the bitmap
+    if ( args[ARG_x1].u_obj == mp_const_none ) {
+        x1 = 0;
+    } else {
+        x1 = mp_obj_get_int(args[ARG_x1].u_obj);
+    }
+    //int16_t y1;
+    if ( args[ARG_y1].u_obj == mp_const_none ) {
+        y1 = 0;
+    } else {
+        y1 = mp_obj_get_int(args[ARG_y1].u_obj);
+    }
+   
+    int16_t x2, y2;
+    // if x2 or y2 is None, then set as the maximum size of the source bitmap
+    if ( args[ARG_x2].u_obj == mp_const_none ) {
+        x2 = source->width-1;
+    } else {
+        x2 = mp_obj_get_int(args[ARG_x2].u_obj);
+    }
+    //int16_t y2;
+    if ( args[ARG_y2].u_obj == mp_const_none ) {
+        y2 = source->height-1;
+    } else {
+        y2 = mp_obj_get_int(args[ARG_y2].u_obj);
+    }
+
+
+    // Check x,y are within self (target) bitmap boundary
+    if ( (x < 0) || (y < 0) || (x >= self->width) || (y >= self->height) ) {
+            mp_raise_ValueError(translate("(x,y): out of range of target bitmap"));
+    }
+    // Check x1,y1,x2,y2 are within source bitmap boundary
+    if ( (x1 < 0) || (x1 >= source->width)  || 
+        (y1 < 0) || (y1 >= source->height) || 
+        (x2 < 0) || (x2 >= source->width)  ||
+        (y2 < 0) || (y2 >= source->height) ) {
+            mp_raise_ValueError(translate("(x1,y1) or (x2,y2): out of range of source bitmap"));
+    }
+    
+    // Ensure x1 < x2 and y1 < y2
+    if (x1 > x2) {
+        int16_t temp=x2;
+        x2=x1;
+        x1=temp;
+    }
+    if (y1 > y2) {
+        int16_t temp=y2;
+        y2=y1;
+        y1=temp;
+    }
+
+    uint32_t skip_index; 
+    bool skip_index_none; // flag whether skip_value was None
+
+    if (args[ARG_skip_index].u_obj == mp_const_none ) {
+        skip_index = 0;
+        skip_index_none = true;
+    } else {
+        skip_index = mp_obj_get_int(args[ARG_skip_index].u_obj);
+        skip_index_none = false;
+    }
```

Comment:
What happens if the bitmaps are different bit depths? AKA, how is it handled if the source bitmap has a wider value range than the destination? The simplest thing would be to raise an error if they differ.

Reply:
I updated to raise an error if the target bitmap (self) has fewer bits per pixel than the source.

#### Future extension
Provide a function to remap the palette of a bitmap, so you could chain a palette remap along with this blit function.